Add fragment overlay class for a list of TriggerPrimitives

philiprodrigues opened the pull request:
In order to save lists of trigger primitives in trigger records, we
need a way to pack them into fragments. This commit provides the
overlay class representing an ad hoc serialization of a list of
trigger primitive objects. (It was decided not to use the
functionality of the serialization package to achieve this, as it
would make fragments containing DS objects look different to fragments
containing data from hardware, and would require offline to depend on
a library to read the MsgPack or JSON format used by the serialization
package).

This commit is just the overlay: code to create Fragment objects and
fill them with TP data is in trigger/FragmentMakers.hpp, along with
code to read back the TPs out of Fragments.

We will eventually need similar overlays for TriggerActivity and TriggerCandidate. If the approach in this PR is approved, I'll work on doing the same for these other classes.

philiprodrigues committed:
Add fragment overlay class for a list of TriggerPrimitives

In order to save lists of trigger primitives in trigger records, we
need a way to pack them into fragments. This commit provides the
overlay class representing an ad hoc serialization of a list of
trigger primitive objects. (It was decided not to use the
functionality of the serialization package to achieve this, as it
would make fragments containing DS objects look different to fragments
containing data from hardware, and would require offline to depend on
a library to read the MsgPack or JSON format used by the serialization
package).

This commit is just the overlay: code to create Fragment objects and
fill them with TP data is in trigger/FragmentMakers.hpp, along with
code to read back the TPs out of Fragments

eflumerf commented on include/dataformats/trigger/TriggerPrimitivesFragment.hpp:
```diff
@@ -0,0 +1,44 @@
+  // the C++ standard, although it happens to work. If we don't want
+  // to use this, we can manually calculate offsets into the structure
+  // to get pointers to the individual TriggerPrimitive structures
+  TriggerPrimitive primitives[];
```
Based on my reading of the standards, this only works because we're "lucky" and using the GCC C++ compiler which supports it as an extension. I think I'd prefer removing this member and providing more C++-compliant methods like `operator[]` and `at(size_t idx) const`

philiprodrigues replied:
This is reasonable (and wasn't as hard as I initially thought). Done in 37528f4

eflumerf commented on include/dataformats/trigger/TriggerPrimitivesFragment.hpp:
```diff
@@ -0,0 +1,44 @@
+
+  uint32_t magic = s_tpf_header_magic;
+  uint32_t version = s_tpf_version; // NOLINT(build/unsigned)
+  uint64_t n_trigger_primitives;    // NOLINT(build/unsigned)
```
Probably should be named `num_trigger_primitives` to avoid confusion with member variable convention.

philiprodrigues replied:
Fixed in 9584987

eflumerf commented on include/dataformats/trigger/TriggerPrimitivesFragment.hpp:
```diff
@@ -0,0 +1,44 @@
+#ifndef DATAFORMATS_INCLUDE_DATAFORMATS_TRIGGER_TRIGGERPRIMITIVESFRAGMENT_HPP_
```
File header comment missing, including copyright info

philiprodrigues replied:
Fixed in 30f2d64

eflumerf commented:
The TriggerPrimitiveFragment should probably have its format documented in a docs/TriggerPrimitiveFragment.md file. The TriggerPrimitive struct can probably just be documented in the same file.

philiprodrigues committed:
Replace flexible array member with explicit at() function

The flexible array member that was being used for the primitives is
not part of the C++ standard (though it happens to work with gcc). So
we replace it with at(i) that returns a reference to the ith primitive

philiprodrigues committed:
Rename n_trigger_primitives to num_trigger_primitives

...to avoid confusion with the special m_ and s_ prefixes for members

philiprodrigues committed:
Add standard file header

philiprodrigues committed:
Apply clang-format

eflumerf committed:
Merge branch 'develop' into philiprodrigues/trigger-fragments

eflumerf committed:
Apply linter suggested fixes

eflumerf reviewed (approved):
LGTM

roland-sipos commented:
@bieryAtFnal , should this be merged?

bieryAtFnal commented:
@roland-sipos , we deferred discussion on this, so I don't believe that it can/should be merged yet.  I recall that another option is to store trigger data objects in pre-defined 'struct-like' formats that wouldn't need serialization code (they would just need overlays).
